### PR TITLE
Added islandora_advanced_search module.

### PR DESCRIPTION
**Pull request by nigelgbanks**
# What does this Pull Request do?

Adds a new optional module that allows for include/exclude facets, collection searching, advanced search, etc.

Most of features are covered in this demo:

image

# How should this be tested?

No configuration is provided as part of the module as it will depend on the users content types, fields etc. The `README.md` documents how it can be configured. 

Install and mess around more or less.

# Additional Notes:

Minimal CSS is provided with the module, the assumption is that will largely be handled by the theme as it's gonna differ from site to site.

No new dependencies are added if you consider the dependencies that `islandora_defaults` has.

# Interested parties
@Islandora/8-x-1-x-committers

### CODE REVIEW

**Comment by dannylamb** [comment]
Niiiiiiiice.  Thx @nigelgbanks !

File: modules/islandora_advanced_search/README.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+- [Documentation](#documentation)
+- [Troubleshooting/Issues](#troubleshootingissues)
+- [Maintainers](#maintainers)
+- [Sponsors](#sponsors)
+- [Development](#development)
+- [License](#license)
+
+## Introduction
+
+This module creates an several blocks to support searching as well as enables the use of Ajax.
```
Comment (ajstanley) [inline modules/islandora_advanced_search/README.md]: Minor typo -

**Review by ajstanley — CHANGES_REQUESTED** [review]
An impressive piece of work!
Other than one tiny typo in the README, I'd suggest moving configuration from Configuration/Islandora to Configuration/search and metadata.

**Comment by nigelgbanks** [comment]
> Other than one tiny typo in the README, I'd suggest moving configuration from Configuration/Islandora to Configuration/search and metadata.

I'll do those both on Monday most likely.

File: modules/islandora_advanced_search/src/Plugin/Block/SearchResultsPagerBlock.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('request_stack')->getMasterRequest(),
```
Comment (jordandukart) [inline modules/islandora_advanced_search/src/Plugin/Block/SearchResultsPagerBlock.php]: As per discussion in Slack this is 7.3 specific feature: https://wiki.php.net/rfc/trailing-comma-function-calls.

File: modules/islandora_advanced_search/src/Plugin/Block/AdvancedSearchBlock.php [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('plugin.manager.search_api.display'),
+      $container->get('form_builder'),
+      $container->get('request_stack')->getMasterRequest(),
```
Comment (jordandukart) [inline modules/islandora_advanced_search/src/Plugin/Block/AdvancedSearchBlock.php]: Same as https://github.com/Islandora/islandora/pull/824/files#r589618221

File: modules/islandora_advanced_search/src/Form/AdvancedSearchForm.php [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+  public function __construct(Request $request) {
+    $this->request = $request;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('request_stack')->getMasterRequest(),
```
Comment (jordandukart) [inline modules/islandora_advanced_search/src/Form/AdvancedSearchForm.php]: Same as https://github.com/Islandora/islandora/pull/824/files#r589618221

File: modules/islandora_advanced_search/src/Controller/AjaxBlocksController.php [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('renderer'),
+      $container->get('path.current'),
+      $container->get('router'),
+      $container->get('path_processor_manager'),
+      $container->get('current_route_match'),
+      $container,
```
Comment (jordandukart) [inline modules/islandora_advanced_search/src/Controller/AjaxBlocksController.php]: Same as https://github.com/Islandora/islandora/pull/824/files#r589618221.

File: modules/islandora_advanced_search/README.md (hunk 1 above)
Comment (nigelgbanks) [reply modules/islandora_advanced_search/README.md]: Should be sorted now.

**Comment by nigelgbanks** [comment]
> An impressive piece of work!
> Other than one tiny typo in the README, I'd suggest moving configuration from Configuration/Islandora to Configuration/search and metadata.

Should be done now.

File: modules/islandora_advanced_search/src/Plugin/Block/SearchResultsPagerBlock.php (hunk 2 above)
Comment (nigelgbanks) [reply modules/islandora_advanced_search/src/Plugin/Block/SearchResultsPagerBlock.php]: Sorted

File: modules/islandora_advanced_search/src/Plugin/Block/AdvancedSearchBlock.php (hunk 3 above)
Comment (nigelgbanks) [reply modules/islandora_advanced_search/src/Plugin/Block/AdvancedSearchBlock.php]: Sorted

File: modules/islandora_advanced_search/src/Form/AdvancedSearchForm.php (hunk 4 above)
Comment (nigelgbanks) [reply modules/islandora_advanced_search/src/Form/AdvancedSearchForm.php]: Sorted

File: modules/islandora_advanced_search/src/Controller/AjaxBlocksController.php (hunk 5 above)
Comment (nigelgbanks) [reply modules/islandora_advanced_search/src/Controller/AjaxBlocksController.php]: Sorted

**Comment by nigelgbanks** [comment]
@jordandukart  I fixed the incompatibilities you highlighted. I tried to find a tool that could automate this check for me, but neither https://vfac.fr/projects/php7compatibility or its latest version from source https://github.com/PHPCompatibility/PHPCompatibility could find the issue. Do you know of a tool that could automate this check? I'd like to integrate something into IDE packaged with Isle https://github.com/Islandora-Devops/isle-buildkit/pull/130 so that it can happen automatically.

**Comment by nigelgbanks** [comment]
Also sorry it took a month to get to this, Isle swallowed almost all of my time during that last month.

**Comment by elizoller** [comment]
@ajstanley are you still looking at this one? or does it need another set of 👀 ?

**Review by ajstanley — APPROVED** [review]

**Comment by ajstanley** [comment]
All looking good.  Merged.

**Comment by kayakr** [comment]
@ajstanley @nigelgbanks Good stuff. However, needs search and replace of `decedent` to `descendant`

**Pull request merged**